base/MAPL_ISO8601_Time.F90

```diff
@@ -0,0 +1,720 @@
+   end interface ISO8601DateTime
+
+   type :: ISO8601Duration
+      type(date_fields) :: date
+      type(time_fields) :: time
+   end type ISO8601Duration
+
+   type :: ISO8601TimeInterval
+      type(ISO8601DateTime) :: start_datetime
+      type(ISO8601DateTime) :: end_datetime
+      integer :: repetitions
+   end type ISO8601TimeInterval
+
+contains
+
+
+! SIMPLE CONSTRUCTORS
+
+   pure function construct_date_fields(fields) result(df)
+      integer, dimension(DIM_DF), intent(in) :: fields
+      type(date_fields) :: df
+      df % year = fields(1)
+      df % month = fields(2)
+      df % day = fields(3)
+      df % valid = is_valid_date(fields)
+   end function construct_date_fields
+
+   pure function construct_time_fields(fields) result(tf)
+      integer, dimension(DIM_TF), intent(in) :: fields
+      type(time_fields) :: tf
+      tf % hour = fields(1)
+      tf % minute = fields(2)
+      tf % second = fields(3)
+      tf % millisecond = fields(4)
+      tf % timezone = fields(5)
+      tf % valid = is_valid_time(fields)
+   end function construct_time_fields
+
+! END SIMPLE CONSTRUCTORS
+
+
+! LOW-LEVEL UTILITY FUNCTIONS FOR STRING PARSERS
+
+   pure function make_char_array(s) result(a)
+      character(len=*), intent(in) :: s
+      character, dimension(len(s)) :: a
+      integer :: i
+      do i=1, len(s)
+         a(i) = s(i:i)
+      end do
+   end function make_char_array
+
+   elemental pure logical function is_digit(c)
+      character, intent(in) :: c
+      integer, parameter :: lb = iachar('0') - 1
+      integer, parameter :: ub = iachar('9') + 1
+      is_digit = iachar(c) > lb .and. iachar(c) < ub
+   end function is_digit
+
+   pure logical function are_digits(s)
+      character(len=*), intent(in) :: s
+      are_digits = all(is_digit(make_char_array(s)))
+   end function are_digits
+
+   pure logical function is_delimited_date(s)
+      character(len=*), intent(in) :: s
+      is_delimited_date = len(s)==10 .and. are_digits(s(1:4)) &
+         .and. s(5:5)==DD .and. are_digits(s(6:7)) .and. s(8:8)==DD &
+         .and. are_digits(s(9:10))
+   end function
+   
+   pure logical function is_undelimited_date(s)
+      character(len=*), intent(in) :: s
+      is_undelimited_date = len(s)==8 .and. are_digits(s)
+   end function is_undelimited_date
+
+   pure logical function is_delimited_time(s)
+      character(len=*), intent(in) :: s
+      
+      select case(len(s))
+         case(10)
+            is_delimited_time = s(1:1)==TP .and. are_digits(s(2:3)) &
+               .and. s(4:4)==TD .and. are_digits(s(5:6)) .and. s(7:7)==TD &
+               .and. are_digits(s(8:9)) .and. s(10:10)==TZ
+         case(14)
+            is_delimited_time = s(1:1)==TP .and. are_digits(s(2:3)) &
+               .and. s(4:4)==TD .and. are_digits(s(5:6)) .and. s(7:7)==TD &
+               .and. are_digits(s(8:9)) .and. s(10:10)=='.' &
+               .and. are_digits(s(11:13)) .and. s(14:14)==TZ
```

Comment:
Enough duplication with the previous block that it suggests introduction of another helper procedure?